### ISSUE TITLE
add_meta_data with unique = true deletes existing meta instead of only adding meta if it doesn't exist

### ISSUE THREAD

**Issue by kkmuffme** [woothemes/woocommerce#48718]
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

`add_meta_data` with unique = true will ALWAYS update the existing meta, instead of only setting the meta if it doesn't exist yet (like how it works everywhere else in the WP eco-system e.g. with wp_cache_add, add_metadata, add_post_meta)

https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/abstracts/abstract-wc-data.php#L443

This should be changed to instead check if the meta exists and if it does to not set the new metadata like it works everywhere else, bc otherwise the "add" function with unique flag is just an alias of "update_meta_data"

### Expected behavior

Don't overwrite existing meta

### Actual behavior

Overwrite existing meta

### Steps to reproduce

```
$product = wc_get_product( 123 );
product->add_meta_data( '_demo_meta', 10, true );
$product->save();

product->add_meta_data( '_demo_meta', 20, true );
$product->save();

var_dump( $product->get_meta( '_demo_meta' ) );
```

You'd expect to get 10, but you'll see 20.

### WordPress Environment

WP 6.5
WC 9.0.1

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by dinhtungdu** [comment]
Hi @kkmuffme, thanks for opening the issue!

> (like how it works everywhere else in the WP eco-system e.g. with wp_cache_add, add_metadata, add_post_meta)

I can see the method isn't aligned with similar one in WordPress Core. But I'm afraid that we can't do different, because it's the public method and has been around since 2016. Changing its behavior will break not only WooCommerce Core but also themes/plugins rely on that method.

Pinging @woocommerce/woo-fse to have a second opinion on this. From my side, I recommending closing this issue with a won't fix label.

**Comment by Aljullu** [comment]
> I can see the method isn't aligned with similar one in WordPress Core. But I'm afraid that we can't do different, because it's the public method and has been around since 2016. Changing its behavior will break not only WooCommerce Core but also themes/plugins rely on that method.

+1 I agree it's confusing that this method behaves differently than similar WP methods, but I don't think we can't change it without considering this a breaking change.

My only idea would be to create a new method (ie: `add_metadata`) which doesn't overwrite the metadata and deprecate the current one.

**Comment by kkmuffme** [comment]
That's what I think makes sense.

Alternatively, just deprecating the $unique param (with a `_deprecated_argument` https://developer.wordpress.org/reference/functions/_deprecated_argument/) and creating a new `add_meta_data_unique` 
Since the current `add_meta_data` with $unique = false behaves as expected, the issue is just with the $unique flag.

This would minimize the amount of work caused by this deprecation.

**Comment by tjcafferkey** [comment]
If we create an `add_meta_data_unique` method won't users using this have the same assumptions unless they have read the implementation?

The fundamental issue is that every other `add_` function in WordPress does not update or overwrite values if they already exist. So deprecating the argument here and creating a new `add_` method would have the same problem unless I am misunderstanding your suggestion @kkmuffme?

**Comment by pmcpinto** [comment]
Thanks for the report, it's really helpful.

I know this has been open for a long time, and I appreciate your patience. We're currently going through some older WooCommerce issues, identifying what still needs fixing, and tidying up what's already fixed or out of scope.

I'd like to get an engineer's eyes on this before we make a call, so I've flagged it for developer feedback. Someone will take a proper look at the repro steps and how it's behaving now.

**Comment by kmanijak** [comment]
I read the issue and all the comments and I agree with other commenters: this cannot be change just like that since that would be a breaking change.

I removed the "Bug" label not to suggest this should be just "fixed".

---

As for next steps, I see three options:

1. Add a new method:
   - either addressing exactly the case mentioned, something like `add_meta_data_if_not_exists` and so we have two methods in parallel. But this introduces more confusion.
   - or more generic method like `add_metadata` and deprecate the current one. This was [suggested by Albert](https://github.com/woocommerce/woocommerce/issues/48718#issuecomment-2298816047) and I'm just reiterating.
2. Add a filter that would override the behaviour on demand. Something like `woocommerce_data_add_meta_data_override_existing` that would allow changing this behaviour in opt-in manner. I would avoid making it generic, probably it would be per key to avoid impacting plugins that could rely on the current behaviour.
3. Do nothing and close the issue. It's out there since 2016 and there's no other reports on it so this is public API.

**Given there's not much activity in the issue I personally vote for option 3.**

**Comment by dinhtungdu** [comment]
> Given there's not much activity in the issue I personally vote for option 3.

As I commented above, I also vote for option 3.

**Comment by vladolaru** [comment]
Thanks for the report and for the follow-up discussion here. I rechecked this against current trunk, and the reported behavior is still how `WC_Data::add_meta_data()` works today.

The relevant implementation still deletes existing metadata for the key before appending the new value when `$unique` is `true`:

- [`WC_Data::add_meta_data()`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/abstracts/abstract-wc-data.php#L489-L508)

That means a second call such as `add_meta_data( '_demo_meta', 20, true )` replaces the previous value. That differs from WordPress Core's `add_post_meta( ..., true )`, where the `unique` flag prevents adding a new row if the key already exists: [WordPress `add_post_meta()` reference](https://developer.wordpress.org/reference/functions/add_post_meta/).

I agree with the earlier comments that changing the existing method in place would be a backward-compatibility break. `WC_Data::add_meta_data()` is public API, and current WooCommerce code uses `$unique = true` as "keep one current value for this key" semantics in several places, for example:

- variation line-item attribute metadata;
- backordered item metadata;
- reduced-stock metadata;
- checkout/order metadata such as `is_vat_exempt`;
- CSV import `_original_id` metadata.

If the method changed to add-only-if-absent, those callers could keep stale metadata instead of replacing it with the current value.

Given that, I think the practical decision is either:

1. Close this as a known compatibility constraint and document that `WC_Data::add_meta_data( ..., true )` means "replace existing values so the key remains unique" in WooCommerce CRUD objects.
2. Add a separate, explicitly named API for add-only-if-absent behavior, with tests for saved objects, unsaved objects, and duplicate existing rows.

Between those, I lean toward option 1 for this issue. The current behavior is confusing compared with WordPress Core naming, but it is long-standing public API, and changing or duplicating the metadata API would add more surface area unless there is a stronger current use case.

**Comment by vladolaru** [comment]
People have voted: closing it!
